semgrep_output_v1.atd

```diff
@@ -846,6 +846,75 @@ type cli_output_extra = {
     ~skipped_rules: skipped_rule list;
 }
 
+(*****************************************************************************)
+(* Matching diagnosis *)
+(*****************************************************************************)
+
+type originating_node_kind = [
+  | Focus
+  | Xpattern
+]
+
+type killing_parent_kind = [
+  | Inside
+  | Negation
+  | Filter of string
+]
+
+(* Instead of serving snippets here, we could just give the locations of
+   the patterns and matches.
+   For convenience when scripting with this in rule generation, we will
+   just get the source text here.
+ *)
+type snippet = {
+  line: int;
+  text: string;
+}
+
+(* a "killing parent" is a parent operator which could have
+   killed the unexpected match along its way to being returned
+   Intuitively, these are all the sites at which the rule could
+   have removed the unexpected match, but didn't.
+   Note that because of the order of operations, this technically
+   means that in the following pattern:
+   all:
+     - pattern: A
+     - not: B
+   the `not` node is a "parent" of the `pattern` node, even though
+   they are siblings in the actual tree. This is because the ranges
+   of the `pattern` are input to the `not` node.
+ *)
+type killing_parent = {
+  killing_parent_kind: killing_parent_kind;
+  snippet: snippet;
+}
+
+(* TODO: this is only completely faithful for search mode rules.
+   this means that if we have an unexpected search mode finding, we
+   should indeed diagnose all the different areas in the rule that
+   could have been responsible for switching the match off.
+   however, for something like a taint mode finding, this is heavily
+   dependent on the structure of the code, so we it's a lot harder.
+   so for taint, secrets, supply chain, this will suggest some reasons
+   but not all, for why a match may be unexpected.
+ *)
+type unexpected_match_diagnosis = {
+  matched_text: snippet;
+  (* information about the originating pattern in the rule file *)
+  (* This is where the unexpected match came from. *)
+  originating_kind: originating_node_kind;
+  originating_text: snippet;
+  killing_parents: killing_parent list;
+}
+
+type unexpected_no_match_diagnosis = todo
+
+type matching_diagnosis = {
```

Comment:
maybe put that first (in ATD the type definition can be in any order, so better put the high-level type first with a comment explaining its use).

Reply:
maybe put an example of use, otherwise it's hard to understand all those type definitions and their use.

Reply:
Please, move this type at the beginning of this section and add some explanations for what a matching diagnosis is.
I still don't understand what it is and also why it's related to 'semgrep test' 
Also what is its relation to the matching explanation? How it differs?

Reply:
oops, i addressed this and lost the changes accidentally in Git. added back now